Libplanet/Blockchain/BlockChain.TxExecution.cs

```diff
@@ -0,0 +1,118 @@
+                .Where(e => e.InputContext.TxId is { })
+                .GroupBy(e => e.InputContext.TxId!.Value);
+            int count = 0;
+            foreach (IGrouping<TxId, ActionEvaluation> txEvals in evaluationsPerTxs)
+            {
+                TxId txid = txEvals.Key;
+                IAccountStateDelta prevStates = txEvals.First().InputContext.PreviousStates;
+                ActionEvaluation evalSum = txEvals.Last();
+                TxExecution txExecution;
+                if (evalSum.Exception is { } e)
+                {
+                    txExecution = new TxFailure(block.Hash, txid, e.InnerException ?? e);
+                }
+                else
+                {
+                    IAccountStateDelta outputStates = evalSum.OutputStates;
+                    txExecution = new TxSuccess(
+                        block.Hash,
+                        txid,
+                        outputStates.GetUpdatedStates(),
+                        outputStates.UpdatedFungibleAssets.ToImmutableDictionary(
+                            kv => kv.Key,
+                            kv => (IImmutableDictionary<Currency, FungibleAssetValue>)kv.Value
+                                .ToImmutableDictionary(
+                                    currency => currency,
+                                    currency => outputStates.GetBalance(kv.Key, currency) -
+                                        prevStates.GetBalance(kv.Key, currency)
+                                )
+                        ),
+                        outputStates.UpdatedFungibleAssets.ToImmutableDictionary(
+                            kv => kv.Key,
+                            kv => (IImmutableDictionary<Currency, FungibleAssetValue>)kv.Value
+                                .ToImmutableDictionary(
+                                    currency => currency,
+                                    currency => outputStates.GetBalance(kv.Key, currency)
+                                )
+                        )
+                    );
+                }
+
+                yield return txExecution;
+                count++;
+            }
+
+            _logger.Verbose(
+                "Prepared " + nameof(TxExecution) +
+                "s for {Txs} transactions within the block #{BlockIndex} {BlockHash}.",
+                count,
+                block.Index,
+                block.Hash
+            );
+        }
+
+        internal void UpdateTxExecutions(IEnumerable<TxExecution> txExecutions)
+        {
+            int count = 0;
+            foreach (TxExecution txExecution in txExecutions)
+            {
+                // Note that there are two overloaded methods of the same name PutTxExecution()
+                // in IStore.  As those two have different signatures, run-time polymorphism
+                // does not work.  Instead, we need the following hard-coded branch:
+                switch (txExecution)
```

Comment:
If only `TxSuccess` and `TxFailure` is allowed as last discussion, how about deal other cases with `default` and throw exception?

Reply:
I left [`Trace.Assert()`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.assert?view=netstandard-2.0) instead: 8272e2c4db35bf3a308bedba5edbd993e74dfdab.